File: ai_fleet/agent/tools/exec.py
"""Shared foundation for the developer-tool registry (port of agent/tools/exec.js).

Every tool in ``tools/`` DELEGATES to a pre-installed standard CLI rather than
re-implementing its behaviour — this module is the one safe door those
delegations go through.

Security posture (see infrastructure-misconfig / secret-leakage checklists):
  - Commands run via ``asyncio.create_subprocess_exec`` with an ARGUMENT ARRAY
    and NEVER a shell string, so tool inputs can never be interpreted as shell
    metacharacters (no command injection).
  - The child inherits the real environment (build tools need $HOME/$PATH/
    $ANDROID_HOME/$JAVA_HOME) but every credential-looking variable is STRIPPED
    first, and any known secret value is REDACTED from returned output — secrets
    never reach the model or child processes.
  - A tool's optional ``dir`` is resolved INSIDE the workspace root; traversal
    ("../etc") is refused.
"""

from __future__ import annotations

import asyncio
import os
import re
import signal
import sys

from ai_fleet.config import CONFIG

DEFAULT_TIMEOUT_SEC = 900
DEFAULT_MAX_OUTPUT_BYTES = 64 * 1024

_MAX_BUFFER_BYTES = 16 * 1024 * 1024


def _to_number(value):
    """JS ``Number(x)`` returning None for NaN/invalid (so ``x or default`` works)."""
    try:
        n = float(value)
    except (TypeError, ValueError):
        return None
    if n != n:  # NaN
        return None
    return int(n) if n == int(n) else n


def tool_limits():
    t = getattr(CONFIG, "TOOLS", None)
    return {
        "timeoutSec": _to_number(getattr(t, "timeoutSec", None)) or DEFAULT_TIMEOUT_SEC,
        "maxOutputBytes": _to_number(getattr(t, "maxOutputBytes", None)) or DEFAULT_MAX_OUTPUT_BYTES,
    }


# Env variable names whose VALUE is a credential — matched case-insensitively.
SECRET_KEY_RE = re.compile(
    r"(TOKEN|SECRET|PASSWORD|PASSWD|PWD|CREDENTIAL|PRIVATE_KEY|ACCESS_KEY|APIKEY|"
    r"API_KEY|_KEY$|SESSION|COOKIE|BEARER|_PAT$|AUTH_|_AUTH$)",
    re.IGNORECASE,
)

# Non-interactive / reproducible flags forced into every tool subprocess.
NONINTERACTIVE_ENV = {
    "GIT_TERMINAL_PROMPT": "0",
    "GCM_INTERACTIVE": "Never",
    "DEBIAN_FRONTEND": "noninteractive",
    "PIP_DISABLE_PIP_VERSION_CHECK": "1",
    "PYTHONUNBUFFERED": "1",
    "npm_config_fund": "false",
    "npm_config_audit": "false",
}


def sanitized_tool_env(base_env=None):
    """Split a base environment into the env a tool subprocess may see and the
    list of secret values to redact from its output. Inherit-then-strip: keep
    every variable a real toolchain needs, drop anything that looks like a
    credential. Returns ``{"env": {...}, "secrets": [...]}``.
    """
    if base_env is None:
        base_env = os.environ
    env = {}
    secrets = []
    for key, value in base_env.items():
        if not isinstance(value, str):
            continue
        if SECRET_KEY_RE.search(key):
            if len(value) >= 4:
                secrets.append(value)
            continue  # stripped — never forwarded to the child
        env[key] = value
    return {"env": {**env, **NONINTERACTIVE_ENV}, "secrets": secrets}


_AUTH_RE = re.compile(
    r"((?:proxy-)?authorization\s*[:=]\s*)(?:bearer|basic|token)?\s*[A-Za-z0-9._~+/=-]{8,}",
    re.IGNORECASE,
)
_KV_SECRET_RE = re.compile(
    r"((?:private-token|x-api-key|api[_-]?key|access[_-]?token|secret|password|passwd|pwd)"
    r"[\"']?\s*[:=]\s*[\"']?)[^\s'\";,&]+",
    re.IGNORECASE,
)
_GH_TOKEN_RE = re.compile(r"\bgh[pousr]_[A-Za-z0-9]{16,}\b")
_GLPAT_RE = re.compile(r"\bglpat-[A-Za-z0-9_-]{16,}\b")
_SK_RE = re.compile(r"\bsk-[A-Za-z0-9]{16,}\b")
_XOX_RE = re.compile(r"\bxox[baprs]-[A-Za-z0-9-]{10,}\b")

REDACTED = "«redacted»"  # «redacted»


def redact_secrets(text, secrets=None, max_bytes=None):
    """Redact known secret values and common credential patterns, then bound length."""
    out = str("" if text is None else text)
    for secret in secrets or []:
        s = str(secret)
        if len(s) >= 4:
            out = out.replace(s, REDACTED)
    out = _AUTH_RE.sub(r"\g<1>" + REDACTED, out)
    out = _KV_SECRET_RE.sub(r"\g<1>" + REDACTED, out)
    out = _GH_TOKEN_RE.sub(REDACTED, out)
    out = _GLPAT_RE.sub(REDACTED, out)
    out = _SK_RE.sub(REDACTED, out)
    out = _XOX_RE.sub(REDACTED, out)
    return truncate(out, max_bytes or tool_limits()["maxOutputBytes"])


def truncate(text, max_bytes):
    """Bound output to ``max_bytes``, keeping the head and tail (errors cluster at the end)."""
    s = str("" if text is None else text)
    if len(s) <= max_bytes:
        return s
    head = int(max_bytes * 0.6)
    tail = max_bytes - head
    dropped = len(s) - max_bytes
    return f"{s[:head]}\n…[truncated {dropped} chars]…\n{s[len(s) - tail:]}"


def resolve_workdir(ctx=None, dir=None):
    """Resolve a tool's working directory INSIDE the workspace root carried on
    ctx. Refuses any ``dir`` that escapes the root (path traversal).
    """
    ctx = ctx or {}
    base = os.path.abspath(ctx.get("cwd") or ctx.get("rootDir") or os.getcwd())
    if not dir:
        return base
    target = os.path.abspath(os.path.join(base, str(dir)))
    relative = os.path.relpath(target, base)
    if relative == ".." or relative.startswith(".." + os.sep) or os.path.isabs(relative):
        raise Exception(f'refusing to operate outside the workspace: "{dir}"')
    return target


def _decode(data) -> str:
    if data is None:
        return ""
    if isinstance(data, (bytes, bytearray)):
        return bytes(data).decode("utf-8", errors="replace")
    return str(data)


async def run_command(command, args=None, opts=None):
    """Run one command with an argument array (never a shell). Resolves with a
    normalized result and NEVER raises on non-zero exit — callers format the
    outcome for the model. Raises only on programmer error (bad arguments).

    Returns a dict: ``{ok, code, signal, stdout, stderr, timedOut, notFound}``.
    """
    if not isinstance(command, str) or not command:
        raise Exception("run_command: command must be a non-empty string")
    if args is None:
        args = []
    if not isinstance(args, list) or any(not isinstance(a, str) or "\0" in a for a in args):
        raise Exception("run_command: args must be an array of strings without null bytes")
    opts = opts or {}
    limits = tool_limits()
    timeout_sec = _to_number(opts.get("timeoutSec")) or limits["timeoutSec"]
    env = opts["env"] if opts.get("env") else sanitized_tool_env()["env"]

    try:
        proc = await asyncio.create_subprocess_exec(
            command,
            *args,
            cwd=opts.get("cwd"),
            env=env,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
            limit=_MAX_BUFFER_BYTES,
        )
    except FileNotFoundError:
        return {"ok": False, "code": 127, "signal": None, "stdout": "", "stderr": "", "timedOut": False, "notFound": True}

    try:
        out, err = await asyncio.wait_for(proc.communicate(), timeout=timeout_sec)
    except asyncio.TimeoutError:
        try:
            proc.kill()
        except ProcessLookupError:
            pass
        try:
            out, err = await proc.communicate()
        except Exception:
            out, err = b"", b""
        return {
            "ok": False,
            "code": None,
            "signal": "SIGTERM",
            "stdout": _decode(out),
            "stderr": _decode(err),
            "timedOut": True,
            "notFound": False,
        }

    rc = proc.returncode
    stdout = _decode(out)
    stderr = _decode(err)
    if rc == 0:
        return {"ok": True, "code": 0, "signal": None, "stdout": stdout, "stderr": stderr, "timedOut": False, "notFound": False}
    if rc is not None and rc < 0:
        try:
            signame = signal.Signals(-rc).name
        except ValueError:
            signame = str(-rc)
        return {"ok": False, "code": None, "signal": signame, "stdout": stdout, "stderr": stderr, "timedOut": False, "notFound": False}
    return {"ok": False, "code": rc, "signal": None, "stdout": stdout, "stderr": stderr, "timedOut": False, "notFound": False}


def platform_cmd(name):
    """Append the platform command-shim suffix (``.cmd``) on Windows for node CLIs."""
    return f"{name}.cmd" if sys.platform.startswith("win") else name


def _step_fn(ctx):
    step = ctx.get("step") if ctx else None
    return step if callable(step) else (lambda *a, **k: None)


async def exec_tool(*, ctx=None, label, command, args, dir=None, timeout_sec=None, not_found_hint=None):
    """The DRY core each executing tool calls: resolve cwd, announce progress,
    run, then return a redacted, bounded, model-readable summary string.
    """
    ctx = ctx or {}
    step = _step_fn(ctx)
    cwd = resolve_workdir(ctx, dir)
    senv = sanitized_tool_env()
    env, secrets = senv["env"], senv["secrets"]
    printable = f"{command} {' '.join(args)}".strip()
    step(f"\U0001f6e0️  {label}: {printable[:120]}")
    result = await run_command(command, args, {"cwd": cwd, "timeoutSec": timeout_sec, "env": env})
    if result["notFound"]:
        hint = not_found_hint or f"Install the {command} CLI (this tool delegates to it) and retry."
        return f"❌ {label}: `{command}` is not installed / not on PATH.\n{hint}"
    return format_result(label=label, command=command, args=args, cwd=cwd, result=result, secrets=secrets)


async def command_exists(command, version_arg="--version"):
    """Whether a CLI is available on PATH (cheap probe via its version flag)."""
    try:
        env = sanitized_tool_env()["env"]
        result = await run_command(command, [version_arg], {"env": env, "timeoutSec": 30})
        return not result["notFound"]
    except Exception:
        return False


async def run_sequence(*, ctx=None, dir=None, steps=None, timeout_sec=None):
    """Run an ordered list of steps in one working directory, stopping at the
    first failure. Returns ``{"ok": bool, "output": str}``.
    """
    ctx = ctx or {}
    steps = steps or []
    cwd = resolve_workdir(ctx, dir)
    senv = sanitized_tool_env()
    env, secrets = senv["env"], senv["secrets"]
    step = _step_fn(ctx)
    chunks = []
    for s in steps:
        step(f"\U0001f6e0️  {s['label']}: {s['command']} {' '.join(s['args'])}"[:140])
        result = await run_command(s["command"], s["args"], {"cwd": cwd, "env": env, "timeoutSec": s.get("timeoutSec") or timeout_sec})
        if result["notFound"]:
            hint = s.get("notFoundHint") or f"Install {s['command']} and retry."
            chunks.append(f"❌ {s['label']}: `{s['command']}` is not installed / not on PATH.\n{hint}")
            return {"ok": False, "output": "\n\n".join(chunks)}
        chunks.append(format_result(label=s["label"], command=s["command"], args=s["args"], cwd=cwd, result=result, secrets=secrets))
        if not result["ok"]:
            return {"ok": False, "output": "\n\n".join(chunks)}
    return {"ok": True, "output": "\n\n".join(chunks)}


def format_result(*, label, command, args, cwd, result, secrets=None):
    """Format a completed command result into the string returned to the model."""
    secrets = secrets or []
    max_output_bytes = tool_limits()["maxOutputBytes"]
    if result["ok"]:
        status = "✅ ok"
    elif result["timedOut"]:
        status = "⏱️ timed out"
    else:
        signal_suffix = f" ({result['signal']})" if result["signal"] else ""
        status = f"❌ exit {result['code']}{signal_suffix}"
    header = f"{status} — {label}\n$ {command} {' '.join(args)}\n(cwd: {cwd})"
    body = "\n".join(
        s for s in ((result["stdout"] or "").strip(), (result["stderr"] or "").strip()) if s
    )
    redacted = redact_secrets(body, secrets, max_output_bytes)
    return f"{header}\n\n{redacted}" if redacted else header


def define_tool(def_, handler):
    """Build a LangChain tool factory with a lazy langchain import (matching the
    registry convention) and a uniform try/except that returns an error string
    instead of raising — one thrown tool must not abort the agent run.

    ``def_`` is a dict ``{"name", "description", "schema"}`` where ``schema`` is a
    pydantic ``BaseModel`` subclass (the JS zod schema, ported to pydantic).
    ``handler`` is ``async (input: dict, ctx: dict) -> str``.
    """

    def factory(ctx=None):
        ctx = ctx or {}
        from langchain_core.tools import StructuredTool

        async def _run(**kwargs):
            try:
                return await handler(kwargs, ctx)
            except Exception as err:
                msg = getattr(err, "message", None) or str(err)
                return f"❌ {def_['name']} failed: {redact_secrets(msg)}"

        return StructuredTool(
            name=def_["name"],
            description=def_["description"],
            args_schema=def_["schema"],
            coroutine=_run,
        )

    return factory

File: packages/shared/src/agent/tools/exec.js
'use strict';

const path = require('path');
const { execFile } = require('child_process');
const { promisify } = require('util');
const { CONFIG } = require('../../config');

const execFileP = promisify(execFile);

/**
 * Shared foundation for the developer-tool registry (docker, environments,
 * build, android, security, quality, codegen, playwright). Every tool in
 * `tools/` DELEGATES to a pre-installed standard CLI rather than re-implementing
 * its behaviour — this module is the one safe door those delegations go through.
 *
 * Security posture (see infrastructure-misconfig / secret-leakage checklists):
 *   - Commands run via execFile with an ARGUMENT ARRAY and NEVER a shell string,
 *     so tool inputs can never be interpreted as shell metacharacters (no
 *     command injection). `shell: false` is the default and is asserted.
 *   - The child inherits the real environment (build tools need $HOME/$PATH/
 *     $ANDROID_HOME/$JAVA_HOME) but every credential-looking variable is
 *     STRIPPED first, and any known secret value is REDACTED from returned
 *     output — secrets never reach the model or child processes.
 *   - A tool's optional `dir` is resolved INSIDE the workspace root; traversal
 *     ("../etc") is refused.
 */

const DEFAULT_TIMEOUT_SEC = 900;
const DEFAULT_MAX_OUTPUT_BYTES = 64 * 1024;

function toolLimits() {
  const t = (CONFIG && CONFIG.TOOLS) || {};
  return {
    timeoutSec: Number(t.timeoutSec) || DEFAULT_TIMEOUT_SEC,
    maxOutputBytes: Number(t.maxOutputBytes) || DEFAULT_MAX_OUTPUT_BYTES,
  };
}

/** Env variable names whose VALUE is a credential — matched case-insensitively. */
const SECRET_KEY_RE =
  /(TOKEN|SECRET|PASSWORD|PASSWD|PWD|CREDENTIAL|PRIVATE_KEY|ACCESS_KEY|APIKEY|API_KEY|_KEY$|SESSION|COOKIE|BEARER|_PAT$|AUTH_|_AUTH$)/i;

/** Non-interactive / reproducible flags forced into every tool subprocess. */
const NONINTERACTIVE_ENV = Object.freeze({
  GIT_TERMINAL_PROMPT: '0',
  GCM_INTERACTIVE: 'Never',
  DEBIAN_FRONTEND: 'noninteractive',
  PIP_DISABLE_PIP_VERSION_CHECK: '1',
  PYTHONUNBUFFERED: '1',
  npm_config_fund: 'false',
  npm_config_audit: 'false',
});

/**
 * Cloud Run exposes the revision service identity through a link-local metadata
 * server. A repository-native test script is untrusted code, so letting that
 * child reach metadata would turn the tester's Firestore/PubSub identity into a
 * credential-exfiltration path. Linux production workers therefore wrap trusted
 * repository commands with a tiny seccomp launcher that permits only AF_UNIX
 * sockets and denies internet/link-local socket creation. Unlike namespace
 * sandboxes, this works without capabilities or user namespaces in Cloud Run.
 * The wrapper is mandatory when `isolateNetwork` is requested; a missing or
 * unsupported sandbox fails closed instead of silently executing with network.
 */
const NETWORK_SANDBOX_COMMAND = 'ai-fleet-network-sandbox';

function networkSandboxInvocation(command, args, opts = {}) {
  if (opts.isolateNetwork !== true) return { command, args };
  const platform = opts.platform || process.platform;
  if (platform !== 'linux') {
    throw new Error('network-isolated commands require a Linux worker');
  }
  return {
    command: opts.networkSandboxCommand || NETWORK_SANDBOX_COMMAND,
    args: [command, ...args],
  };
}

/**
 * Split a base environment into the env a tool subprocess may see and the list
 * of secret values to redact from its output. Inherit-then-strip: keep every
 * variable a real toolchain needs, drop anything that looks like a credential.
 * @param {NodeJS.ProcessEnv} [baseEnv]
 * @returns {{ env: Record<string,string>, secrets: string[] }}
 */
function sanitizedToolEnv(baseEnv = process.env) {
  const env = {};
  const secrets = [];
  for (const [key, value] of Object.entries(baseEnv)) {
    if (typeof value !== 'string') continue;
    if (SECRET_KEY_RE.test(key)) {
      if (value.length >= 4) secrets.push(value);
      continue; // stripped — never forwarded to the child
    }
    env[key] = value;
  }
  return { env: { ...env, ...NONINTERACTIVE_ENV }, secrets };
}

/**
 * Redact known secret values and common credential patterns, then bound length.
 * @param {unknown} text
 * @param {string[]} [secrets]  literal secret values to blank out
 * @param {number} [maxBytes]
 */
function redactSecrets(text, secrets = [], maxBytes) {
  let out = String(text == null ? '' : text);
  for (const secret of secrets) {
    const s = String(secret);
    if (s.length >= 4) out = out.split(s).join('«redacted»');
  }
  out = out
    .replace(/((?:proxy-)?authorization\s*[:=]\s*)(?:bearer|basic|token)?\s*[A-Za-z0-9._~+/=-]{8,}/gi, '$1«redacted»')
    .replace(/((?:private-token|x-api-key|api[_-]?key|access[_-]?token|secret|password|passwd|pwd)["']?\s*[:=]\s*["']?)[^\s'";,&]+/gi, '$1«redacted»')
    .replace(/\bgh[pousr]_[A-Za-z0-9]{16,}\b/g, '«redacted»')
    .replace(/\bglpat-[A-Za-z0-9_-]{16,}\b/g, '«redacted»')
    .replace(/\bsk-[A-Za-z0-9]{16,}\b/g, '«redacted»')
    .replace(/\bxox[baprs]-[A-Za-z0-9-]{10,}\b/g, '«redacted»');
  return truncate(out, maxBytes || toolLimits().maxOutputBytes);
}

/** Bound output to `maxBytes`, keeping the head and tail (errors cluster at the end). */
function truncate(text, maxBytes) {
  const s = String(text == null ? '' : text);
  if (s.length <= maxBytes) return s;
  const head = Math.floor(maxBytes * 0.6);
  const tail = maxBytes - head;
  const dropped = s.length - maxBytes;
  return `${s.slice(0, head)}\n…[truncated ${dropped} chars]…\n${s.slice(s.length - tail)}`;
}

/**
 * Resolve a tool's working directory INSIDE the workspace root carried on ctx.
 * Refuses any `dir` that escapes the root (path traversal).
 * @param {{ cwd?: string, rootDir?: string }} ctx
 * @param {string} [dir] optional subdirectory relative to the root
 * @returns {string} absolute path inside the root
 */
function resolveWorkdir(ctx = {}, dir) {
  const base = path.resolve(ctx.cwd || ctx.rootDir || process.cwd());
  if (!dir) return base;
  const target = path.resolve(base, String(dir));
  const relative = path.relative(base, target);
  if (relative === '..' || relative.startsWith(`..${path.sep}`) || path.isAbsolute(relative)) {
    throw new Error(`refusing to operate outside the workspace: "${dir}"`);
  }
  return target;
}

/**
 * Run one command with an argument array (never a shell). Resolves with a
 * normalized result and NEVER rejects on non-zero exit — callers format the
 * outcome for the model. Rejects only on programmer error (bad arguments).
 * @returns {Promise<{ ok:boolean, code:(number|null), signal:(string|null),
 *   stdout:string, stderr:string, timedOut:boolean, notFound:boolean }>}
 */
async function runCommand(command, args = [], opts = {}) {
  if (typeof command !== 'string' || !command) throw new Error('runCommand: command must be a non-empty string');
  if (!Array.isArray(args) || args.some((a) => typeof a !== 'string' || a.includes('\0'))) {
    throw new Error('runCommand: args must be an array of strings without null bytes');
  }
  const limits = toolLimits();
  const timeoutSec = Number(opts.timeoutSec) || limits.timeoutSec;
  const { env } = opts.env ? { env: opts.env } : sanitizedToolEnv();
  const invocation = networkSandboxInvocation(command, args, opts);
  try {
    const { stdout, stderr } = await execFileP(invocation.command, invocation.args, {
      cwd: opts.cwd,
      env,
      shell: false,
      timeout: timeoutSec * 1000,
      maxBuffer: 16 * 1024 * 1024,
      windowsHide: true,
    });
    return { ok: true, code: 0, signal: null, stdout, stderr, timedOut: false, notFound: false };
  } catch (err) {
    if (err && err.code === 'ENOENT') {
      return { ok: false, code: 127, signal: null, stdout: '', stderr: '', timedOut: false, notFound: true };
    }
    return {
      ok: false,
      code: typeof err.code === 'number' ? err.code : null,
      signal: err.signal || null,
      stdout: err.stdout ? String(err.stdout) : '',
      stderr: err.stderr ? String(err.stderr) : String(err.message || err),
      timedOut: Boolean(err.killed) && err.signal === 'SIGTERM',
      notFound: false,
    };
  }
}

/** Append the platform command-shim suffix (`.cmd`) on Windows for node CLIs. */
function platformCmd(name) {
  return process.platform === 'win32' ? `${name}.cmd` : name;
}

/**
 * The DRY core each executing tool calls: resolve cwd, announce progress, run,
 * then return a redacted, bounded, model-readable summary string.
 * @param {object} params
 * @param {object} params.ctx     tool ctx ({ cwd|rootDir, step })
 * @param {string} params.label   short human label for the step line
 * @param {string} params.command executable name
 * @param {string[]} params.args  argument array
 * @param {string} [params.dir]   workspace-relative working directory
 * @param {number} [params.timeoutSec]
 * @param {string} [params.notFoundHint] guidance shown when the CLI is absent
 * @returns {Promise<string>}
 */
async function execTool({ ctx = {}, label, command, args, dir, timeoutSec, notFoundHint }) {
  const step = typeof ctx.step === 'function' ? ctx.step : () => {};
  const cwd = resolveWorkdir(ctx, dir);
  const { env, secrets } = sanitizedToolEnv();
  const printable = `${command} ${args.join(' ')}`.trim();
  step(`🛠️  ${label}: ${printable.slice(0, 120)}`);
  const result = await runCommand(command, args, {
    cwd,
    timeoutSec,
    env,
    isolateNetwork: ctx.isolateNetwork === true,
  });
  if (result.notFound) {
    return `❌ ${label}: \`${command}\` is not installed / not on PATH.\n${notFoundHint || `Install the ${command} CLI (this tool delegates to it) and retry.`}`;
  }
  return formatResult({ label, command, args, cwd, result, secrets });
}

/**
 * Whether a CLI is available on PATH (cheap probe via its version flag).
 * @param {string} command
 * @param {string} [versionArg]  e.g. '--version' (default) or 'version' for `go`
 * @returns {Promise<boolean>}
 */
async function commandExists(command, versionArg = '--version') {
  try {
    const { env } = sanitizedToolEnv();
    const result = await runCommand(command, [versionArg], { env, timeoutSec: 30 });
    return !result.notFound;
  } catch (_) {
    return false;
  }
}

/**
 * Run an ordered list of steps in one working directory, stopping at the first
 * failure. Used by orchestration tools (env/build/android) that chain several
 * standard CLIs. Returns the concatenated, redacted output and overall success.
 * @param {object} params
 * @param {object} params.ctx
 * @param {string} [params.dir]
 * @param {Array<{label:string,command:string,args:string[],timeoutSec?:number,notFoundHint?:string}>} params.steps
 * @param {number} [params.timeoutSec]
 * @returns {Promise<{ ok:boolean, output:string }>}
 */
async function runSequence({ ctx = {}, dir, steps = [], timeoutSec }) {
  const cwd = resolveWorkdir(ctx, dir);
  const { env, secrets } = sanitizedToolEnv();
  const step = typeof ctx.step === 'function' ? ctx.step : () => {};
  const chunks = [];
  for (const s of steps) {
    step(`🛠️  ${s.label}: ${s.command} ${s.args.join(' ')}`.slice(0, 140));
    const result = await runCommand(s.command, s.args, {
      cwd,
      env,
      timeoutSec: s.timeoutSec || timeoutSec,
      isolateNetwork: ctx.isolateNetwork === true,
    });
    if (result.notFound) {
      chunks.push(`❌ ${s.label}: \`${s.command}\` is not installed / not on PATH.\n${s.notFoundHint || `Install ${s.command} and retry.`}`);
      return { ok: false, output: chunks.join('\n\n') };
    }
    chunks.push(formatResult({ label: s.label, command: s.command, args: s.args, cwd, result, secrets }));
    if (!result.ok) return { ok: false, output: chunks.join('\n\n') };
  }
  return { ok: true, output: chunks.join('\n\n') };
}

/** Format a completed command result into the string returned to the model. */
function formatResult({ label, command, args, cwd, result, secrets = [] }) {
  const { maxOutputBytes } = toolLimits();
  const status = result.ok ? '✅ ok' : result.timedOut ? '⏱️ timed out' : `❌ exit ${result.code}${result.signal ? ` (${result.signal})` : ''}`;
  const header = `${status} — ${label}\n$ ${command} ${args.join(' ')}\n(cwd: ${cwd})`;
  const body = [result.stdout, result.stderr].map((s) => (s || '').trim()).filter(Boolean).join('\n');
  const redacted = redactSecrets(body, secrets, maxOutputBytes);
  return redacted ? `${header}\n\n${redacted}` : header;
}

/**
 * Build a LangChain tool factory with lazy langchain/zod requires (matching the
 * registry convention) and a uniform try/catch that returns an error string
 * instead of throwing — one thrown tool must not abort the agent run.
 * @param {{ name:string, description:string, schema:(z:any)=>any }} def
 * @param {(input:object, ctx:object)=>Promise<string>|string} handler
 * @returns {(ctx?:object)=>object} factory usable in the FACTORIES registry
 */
function defineTool(def, handler) {
  return (ctx = {}) => {
    const { tool } = require('@langchain/core/tools');
    const { z } = require('zod');
    return tool(
      async (input) => {
        try {
          return await handler(input || {}, ctx);
        } catch (err) {
          const msg = err && err.message ? err.message : String(err);
          return `❌ ${def.name} failed: ${redactSecrets(msg)}`;
        }
      },
      { name: def.name, description: def.description, schema: def.schema(z) }
    );
  };
}

module.exports = {
  DEFAULT_TIMEOUT_SEC,
  DEFAULT_MAX_OUTPUT_BYTES,
  SECRET_KEY_RE,
  sanitizedToolEnv,
  redactSecrets,
  truncate,
  resolveWorkdir,
  runCommand,
  networkSandboxInvocation,
  NETWORK_SANDBOX_COMMAND,
  platformCmd,
  commandExists,
  runSequence,
  execTool,
  formatResult,
  defineTool,
  toolLimits,
};
